### PR TITLE
 Add a handshake parameter to wrapSocket() to allow it to work on an already-connected socket.

### PR DESCRIPTION
**Pull request by wiml**
(This is a new version of [pull request 2513](https://github.com/Araq/Nim/pull/2513) so it can pull from a different branch)

Some protocols involve an initial cleartext exchange before the SSL setup starts, which means having to wrap an already-connected socket. This patch adds an argument to wrapSocket() telling it to go ahead and initiate the SSL negotiation (and which side of the negotiation it's on, since SSL is asymmetric wrt the client and server).

I've tested this for both `net` and `asyncnet` but only from the client side.

### CODE REVIEW

File: lib/pure/net.nim [hunk 1]
```diff
@@ -249,6 +258,24 @@ when defined(ssl):
     if SSLSetFd(socket.sslHandle, socket.fd) != 1:
       raiseSSLError()
 
+  proc wrapSocket*(ctx: SSLContext, socket: Socket, handshake: SslHandshakeType) =
```
Comment (dom96) [inline lib/pure/net.nim]: Please rename this to ``wrapConnectedSocket``.

File: lib/pure/asyncnet.nim [hunk 2]
```diff
@@ -434,6 +434,15 @@ when defined(ssl):
     socket.bioOut = bioNew(bio_s_mem())
     sslSetBio(socket.sslHandle, socket.bioIn, socket.bioOut)
 
+  proc wrapSocket*(ctx: SslContext, socket: AsyncSocket, handshake: SslHandshakeType) =
```
Comment (dom96) [inline lib/pure/asyncnet.nim]: Add some docs and rename it to ``wrapConnectedSocket`` please.

**Comment by Varriount** [comment]
@wiml Any progress on this?

**Comment by Araq** [comment]
@dom96 pull it and take over please.

**Pull request merged**